### ISSUE TITLE
Pedestrian isochrones have unnatural extensions

### ISSUE THREAD

**Issue by nkoren** [valhalla/valhalla#3214]
Here's a pedestrian isochrone with 15-minute contours:

![image](https://user-images.githubusercontent.com/1158752/125775390-c8c0de8e-5925-4701-8cc6-1443c1b78945.png)

See that long "tail" off to the right, following the river? That whole thing shouldn't exist. There are also somewhat less dramatic "tails" on the left, also following the river, which shouldn't exist.

No idea what the underlying mechanism for this would be, but figured it was worth tracking!

**Comment by dnesbitt61** [comment]
These are likely correct. There are ferries that allow reaching the  "tail" to the right from the origin within an hour. You would need to add a costing option to change the ferry use if you want to exclude ferries from consideration.

**Comment by TimMcCauley** [comment]
To give you an example @nkoren  - https://www.openstreetmap.org/way/459235633#map=14/51.4966/-0.0413

**Comment by nkoren** [comment]
Ah, but since this is pedestrian routing, it doesn't include public modes
like metro or bus. Why would it include ferries?

**Comment by kevinkreiser** [comment]
its a bit of an implementation detail but pedestrian and all other modes include ferries without schedule information. bus and train are not included by default we require scheduling info for those. the main reason is because there are places that you just cant reach without takign a ferry and so practically we have to allow them even if we dont have acurate schedule information for them. on the contrary there arent really any places for which you would *need* to include bus and train modes to be able to reach them

**Comment by nkoren** [comment]
I'm definitely not sure that logic is universal enough to be baked into the isochrone calculations in that way. There are many places (as here in London, for example), where the ferries are low-frequency services which would not contribute meaningfully to the walkshed. There are other places where land is inaccessible by walking, but the connection pedestrians would make are much more often by train than by ferry -- Hong Kong to Kowloon, for example -- so privileging the ferries over other modes seems arbitrary. (Although I see that in the latter example, there's no routing, because the ferries going from Hong Kong are not *river* ferries.

In testing various connections, I've also noticed something odd, which is perhaps an unrelated issue. Here's the pedestrian accessibility of New York, starting from a location about 300m east of a ferry terminal (these are 15-minute contours):

![image](https://user-images.githubusercontent.com/1158752/125795652-2a9bd465-1c29-4bea-8958-c7198ea75f00.png)

Looks fine, and the ferries are ignored. Now, we move 300m west, so we're directly at the ferry terminal:

![image](https://user-images.githubusercontent.com/1158752/125795195-deae5351-28f1-4dee-8fb0-1834d4acf09d.png)

Now, vast tracts of New Jersey are accessible -- and the contours in New York die after 30 minutes.

**Comment by nilsnolde** [comment]
for your first paragraph: please do note that there is a `use_ferry` costing option (as @dnesbitt61 mentioned): https://github.com/valhalla/valhalla/blob/master/docs/api/turn-by-turn/api-reference.md#pedestrian-costing-options

still, it _would_ take a ferry if that's the only thing available to get to the destination, as the description says (that's why it's not arbitrary or elevating ferries over road/rail modes, it's just necessary).

your other issue: I seem to remember (but can't prove it with code atm) that ferries are pretty costly for the algorithm. it seems a little over the top in this example, but maybe your first location was (in connectivity terms) pretty far away even though it's only 300 m as-the-crow-flies.

**Comment by nkoren** [comment]
Okay, will take note of `use_ferry`! However for the second issue, I think there's something definitely suspicious going on. Here's a zoomed-in detail of the origin point where we *don't* take any ferries (actually at 190m now). You can see it's a straight shot to the ferry terminal, so shouldn't be a costly walk at all:

![image](https://user-images.githubusercontent.com/1158752/125799719-4d74c055-400f-42f3-a07d-0535eccd9945.png)

Move 10m down the road, and we *do* take the ferries:

![image](https://user-images.githubusercontent.com/1158752/125799949-4db6ac21-0a22-4d8b-85cc-5c34ff6c8861.png)

More disconcertingly, however, we also suddenly lose the 45-60 minute contours for Manhattan:

![image](https://user-images.githubusercontent.com/1158752/125800085-21116c56-f815-4cd6-8744-13de24e401c6.png)

The dividing line between the two states actually seems to be about 180m as the crow flies (and the walk-path isn't much different than crow-flies, in this case).

At about 40m distance, there's another jump, and the 30-45 minute contours in Manhattan also disappear.

**Comment by nilsnolde** [comment]
huh, curious. thanks for the detailed pics & explanation. I just happen to be knee-deep in the isochrone stuff, so I'll try that one one of these days.

**Comment by kevinkreiser** [comment]
@nilsnolde is hitting on an important point. the isochrone still uses the costing model to determine what edges to traverse or are worth traversing, that is, the expansion uses the least cost where cost is a function of time but also of desirability of an edge  in the graph. ferries, as you have noted are somewhat undesirable so they have a penalty associated with them. the closer you are to one the more likely the penalty is overcome before you run out of time.

**Comment by nilsnolde** [comment]
right, you might've juuuust met that limit there @nkoren with your 10 m variation.

**Comment by nilsnolde** [comment]
try an isodistance (`distance` instead of `time`), that should get you across the river (almost) no matter the cost, although there's a check for that too, so you can't jump on an cross-atlantic ferry accidentally ;)

**Comment by kevinkreiser** [comment]
> More disconcertingly, however, we also suddenly lose the 45-60 minute contours for Manhattan

@nkoren did you look at what was actually returned? since those contours would be disjoint its possibly that its a rendering artifact (inside out polygon with wrong winding possibly?) or its also possible that  you are using one of the optional isochrone parameters that removes some contours depending on their relative sizes (denoise parameter)
